### PR TITLE
fix(migrate): getDbInfo `TypeError: Cannot read property 'url' of undefined`

### LINKED ISSUE #9518: `db push` with empty `schema.prisma`: `Error: TypeError: Cannot read properties of undefined (reading 'url')`
Running `db push` with an empty schema file leads to this suboptimal error message:

```
λ npx prisma db push
Need to install the following packages:
  prisma
Ok to proceed? (y) y
  prisma:loadEnv skipping package.json at C:\Users\Jan\Documents\throwaway\9184repro\foo\package.json +0ms
  prisma:loadEnv project root found at C:\Users\Jan\Documents\throwaway\9184repro\package.json +4ms
  prisma:tryLoadEnv Environment variables not found at null +0ms
  prisma:tryLoadEnv Environment variables not found at undefined +3ms
  prisma:tryLoadEnv No Environment variables loaded +9ms
  prisma:engines binaries to download libquery-engine, migration-engine, introspection-engine, prisma-fmt +0ms
Prisma schema loaded from schema.prisma
  prisma:getConfig Using CLI Query Engine (Node-API Library) at: C:\Users\Jan\AppData\Local\npm-cache\_npx\2778af9cee32ff87\node_modules\@prisma\engines\query_engine-windows.dll.node +0ms
Error: TypeError: Cannot read properties of undefined (reading 'url')
    at getDbInfo (C:\Users\Jan\AppData\Local\npm-cache\_npx\2778af9cee32ff87\node_modules\prisma\build\index.js:62954:34)
    at async printDatasource (C:\Users\Jan\AppData\Local\npm-cache\_npx\2778af9cee32ff87\node_modules\prisma\build\index.js:64502:20)
    at async _DbPush.parse (C:\Users\Jan\AppData\Local\npm-cache\_npx\2778af9cee32ff87\node_modules\prisma\build\index.js:66357:7)
    at async main (C:\Users\Jan\AppData\Local\npm-cache\_npx\2778af9cee32ff87\node_modules\prisma\build\index.js:112748:18)
```

Found via https://github.com/prisma/prisma/issues/9517

### PR DESCRIPTION
**Pull request by Jolg42**
Closes https://github.com/prisma/prisma/issues/9518

### CODE REVIEW

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 1]
```diff
@@ -820,6 +820,33 @@ describeIf(process.platform !== 'win32' && !isMacOrWindowsCI)('MongoDB', () => {
     expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
   })
 
+  // In this case it should not and the line `Datasource "x"` not be printed
```
Comment (janpio) [inline packages/migrate/src/__tests__/DbPull.test.ts:823]: What does "it should not" refer to?

File: packages/migrate/src/__tests__/fixtures/schema-only-mongodb/only-generator/schema.prisma [hunk 2]
```diff
@@ -0,0 +1,4 @@
+generator client {
+            provider        = "prisma-client-js"
+            previewFeatures = ["mongoDb"]
```
Comment (janpio) [inline packages/migrate/src/__tests__/fixtures/schema-only-mongodb/only-generator/schema.prisma:3]:
```suggestion
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb"]
```

Comment (Jolg42) [reply packages/migrate/src/__tests__/fixtures/schema-only-mongodb/only-generator/schema.prisma]: Looks like my default spacing was still changed from last time I did a repro on prisma-fmt auto format 😅 
Fixed

File: packages/migrate/src/__tests__/DbPull.test.ts (hunk 1 above)
Comment (Jolg42) [inline packages/migrate/src/__tests__/DbPull.test.ts:823]:
```suggestion
  // In this case it should not error and the line `Datasource "x"` not be printed
```

Comment (Jolg42) [reply packages/migrate/src/__tests__/DbPull.test.ts:823]: error, nice catch

File: packages/migrate/src/commands/DbDrop.ts [hunk 3]
```diff
@@ -75,6 +75,11 @@ ${chalk.bold('Examples')}
     await printDatasource(schemaPath)
 
     const dbInfo = await getDbInfo(schemaPath)
+    if (!dbInfo.url) {
+      // TODO better error
+      throw new Error('Connection url is undefined.')
+    }
```
Comment (Jolg42) [inline packages/migrate/src/commands/DbDrop.ts:81]: `db drop` is not part of public API, we can do that later when we work on it

**Review by millsp — APPROVED** [review]

File: packages/migrate/src/__tests__/DbPull.test.ts [hunk 4]
```diff
@@ -820,6 +820,33 @@ describeIf(process.platform !== 'win32' && !isMacOrWindowsCI)('MongoDB', () => {
     expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
   })
 
+  // In this case it should not error and the line `Datasource "x"` not be printed
+  test('introspection --url - only generator defined', async () => {
+    ctx.fixture('schema-only-mongodb/only-generator')
+    const introspect = new DbPull()
```
[the remark this answers is not in the document]
Comment (Jolg42) [reply packages/migrate/src/__tests__/DbPull.test.ts:826]: The root cause is the same, (same code path used -> getDbInfo)

[the remark this answers is not in the document]
Comment (millsp) [reply packages/migrate/src/__tests__/DbPull.test.ts:826]: Answer: Yes.

**Pull request merged**